cwl/src/main/scala/cwl/preprocessor/CwlPreProcessor.scala

```diff
@@ -0,0 +1,272 @@
+    for {
+      // parse the file containing the reference
+      parsed <- saladAndParse(cwlReference.file)
+      // Get a Map[CwlReference, Json] from the parsed file. If the file is a JSON object and only contains one node, the map will only have 1 element 
+      newUnProcessedReferences = mapIdToContent(parsed).toMap
+      // The reference json in the file
+      referenceJson <- newUnProcessedReferences
+        .collectFirst({ case (ref, json) if ref.pointer == cwlReference.pointer => json })
+        .toParse(s"Cannot find a tool or workflow with ID ${cwlReference.fullReference} in file ${cwlReference.file.pathAsString}")
+      // Process the reference json
+      processed <- flattenJson(referenceJson, newUnProcessedReferences ++ unProcessedReferences, processedReferences, breadCrumbs + cwlReference)
+    } yield processed
+  }
+
+  /**
+    * Given a reference from a step's run field, flattens it and return it
+    * @param unProcessedReferences references that have been parsed and saladed (we have the json), but not flattened yet.
+    * @param checkedProcessedReferences references that are fully processed
+    * @param cwlReference reference being processed
+    * @return a new ProcessedReferences Map including this cwlReference processed along with all the dependencies
+    *         that might have been processed recursively.
+    */
+  private def processCwlRunReference(unProcessedReferences: UnProcessedReferences,
+                                     breadCrumbs: Set[CwlReference])
+                                    (checkedProcessedReferences: Parse[ProcessedReferences],
+                                     cwlReference: CwlReference): Parse[ProcessedReferences] = {
+    checkedProcessedReferences flatMap { processedReferences =>
+      // If the reference has already been processed, no need to do anything
+      if (processedReferences.contains(cwlReference)) processedReferences.validParse
+      // If the reference is in the bread crumbs it means we circled back to it: fail the pre-processing
+      else if (breadCrumbs.contains(cwlReference)) s"Found a circular dependency on $cwlReference".invalidParse
+      else {
+        // Otherwise let's see if we already have the json for it or if we need to process the file
+        val result: Parse[ProcessedJsonAndDependencies] = unProcessedReferences.get(cwlReference) match {
+          case Some(unProcessedReferenceJson) =>
+            // Found the json in the unprocessed map, no need to reparse the file, just flatten this json
+            flattenJson(unProcessedReferenceJson, unProcessedReferences, processedReferences, breadCrumbs)
+          case None =>
+            // This is the first time we're seeing this reference, we need to parse its file and flatten it
+            flattenCwlReference(cwlReference, unProcessedReferences, processedReferences, breadCrumbs)
+        }
+
+        result map {
```

Comment:
ToL feels like this `processCwlRunReference` function could be structured as a forcomp